src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -94,13 +84,122 @@ module Writer = struct
         let%bind _ = Pipe.downstream_flushed t.root_pipe in
         Deferred.unit )
 
-  let close t =
+  let rec close : type a. a t -> unit =
+   fun t ->
     guard_already_closed t (fun () ->
         Pipe.close t.root_pipe ;
-        Int.Table.iter t.pipes ~f:(fun w -> Pipe.close w) ;
+        Int.Table.iter t.pipes ~f:(function
+          | Leaf writer -> Pipe.close writer
+          | Node {child_pipe; _} -> Option.iter (Ivar.peek child_pipe) ~f:close ) ;
         Int.Table.clear t.pipes )
 end
 
+let create a =
+  let root_r, root_w = Pipe.create () in
+  let t =
+    {root_pipe= root_w; cache= a; reader_id= 0; pipes= Int.Table.create ()}
+  in
+  let downstream_flushed_v : unit Ivar.t ref = ref @@ Ivar.create () in
+  let consumer =
+    Pipe.add_consumer root_r ~downstream_flushed:(fun () ->
+        let%map () = Ivar.read !downstream_flushed_v in
+        (* Sub-pipes are never closed without closing the master pipe. *)
+        `Ok )
+  in
+  don't_wait_for
+    (Pipe.iter ~flushed:(Consumer consumer) root_r ~f:(fun v ->
+         t.cache <- v ;
+         downstream_flushed_v := Ivar.create () ;
+         let inner_pipes = Int.Table.data t.pipes in
+         let%bind () =
+           Deferred.List.iter ~how:`Parallel inner_pipes
+             ~f:(function
+             | Leaf writer -> Pipe.write writer v
+             | Node {write; _} -> write v )
+         in
+         Pipe.Consumer.values_sent_downstream consumer ;
+         let%bind () =
+           Deferred.List.iter ~how:`Parallel inner_pipes
+             ~f:(function
+             | Leaf writer -> Deferred.ignore @@ Pipe.downstream_flushed writer
+             | Node _ ->
+                 Deferred.unit
+                 (* A Broadcast pipe would get it's write determined after all of it's readers are consumed *) )
```

Comment:
This comment is confusing and I was sure you had done it wrong until I read further. I see now that we're relying on the `write` function inside the `Node` blocking until the downstream consumers are done. Can you make it clear that `write` always calls `Broadcast_pipe.Writer.write`?